engine/table/src/main/java/io/deephaven/engine/table/impl/ReverseOperation.java

```diff
@@ -149,44 +151,49 @@ private void onUpdate(final TableUpdate upstream) {
         final long newShift =
                 (parentRowSet.lastRowKey() > pivotPoint) ? computePivot(parentRowSet.lastRowKey()) - pivotPoint : 0;
         if (upstream.shifted().nonempty() || newShift > 0) {
-            long watermarkKey = 0;
-            final RowSetShiftData.Builder oShiftedBuilder = new RowSetShiftData.Builder();
-
-            // Bounds seem weird because we might need to shift all keys outside of shifts too.
-            for (int idx = upstream.shifted().size(); idx >= 0; --idx) {
-                final long nextShiftEnd;
-                final long nextShiftStart;
-                final long nextShiftDelta;
-                if (idx == 0) {
-                    nextShiftStart = nextShiftEnd = pivotPoint + 1;
-                    nextShiftDelta = 0;
-                } else {
-                    // Note: begin/end flip responsibilities in the transformation
-                    nextShiftDelta = -upstream.shifted().getShiftDelta(idx - 1);
-                    final long minStart = Math.max(-nextShiftDelta - newShift, 0);
-                    nextShiftStart = Math.max(minStart, transform(upstream.shifted().getEndRange(idx - 1)));
-                    nextShiftEnd = transform(upstream.shifted().getBeginRange(idx - 1));
-                    if (nextShiftEnd < nextShiftStart) {
-                        continue;
+            // Only compute downstream shifts if there are retained rows to shift
+            if (resultRowSet.isEmpty()) {
```

Comment:
You could also skip this if the prev rowset is empty (since shifts affect rows that exist in both key spaces)